### PR TITLE
Replace the getSearchablePages hook in the news bundle

### PR DESCRIPTION
**Pull request by fritzmg**
The `getSearchablePages` hook has been deprecated and will be removed in Contao 5.0. This PR will instead implement a `contao.sitemap` event for news, faqs, newsletters and events.

### CODE REVIEW

**Comment by Toflar** [comment]
Let's remove the hook and everything associated with it too then :)

**Comment by fritzmg** [comment]
Yes, though for better PR readability I prefer to separate these tasks. First remove the self deprecating implementations, then remove the hook and everything related once nothing in the core bundles uses it anymore.

**Comment by fritzmg** [comment]
Let's review and merge this first, before I do the other bundles. Failing tests are unrelated.

**Review by Toflar — DISMISSED** [review]

File: news-bundle/src/EventListener/SitemapListener.php [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+use Contao\CoreBundle\Event\SitemapEvent;
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Contao\News;
+use Contao\NewsArchiveModel;
+use Contao\NewsModel;
+use Contao\PageModel;
+use Terminal42\ServiceAnnotationBundle\Annotation\ServiceTag;
+
+/**
+ * @ServiceTag("kernel.event_listener", event=ContaoCoreEvents::SITEMAP)
```
Comment (aschempp) [inline news-bundle/src/EventListener/SitemapListener.php]: I think we're not using annotations but define our service tags inside the `.yml` file.
(also, we _should_ use PHP8 attributes for Contao 5 😆)

File: news-bundle/src/EventListener/SitemapListener.php [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        $newsArchiveModelAdapter = $this->framework->getAdapter(NewsArchiveModel::class);
+
+        // Get all news archives
+        $archives = $newsArchiveModelAdapter->findAll();
+
+        if (null === $archives) {
+            return;
+        }
+
+        /** @var PageModel $pageModelAdapter */
```
Comment (aschempp) [inline news-bundle/src/EventListener/SitemapListener.php]: No need to declare the types, this should be automatically inferred in our latest return type comments.

File: news-bundle/src/EventListener/SitemapListener.php [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+        /** @var NewsArchiveModel $archive */
+        foreach ($archives as $archive) {
+            // Skip archives without a target page
+            if (empty($archive->jumpTo)) {
+                continue;
+            }
+
+            // Skip archive if protected and not accessible
+            if (!$controllerAdapter->isVisibleElement($archive)) {
```
Comment (aschempp) [inline news-bundle/src/EventListener/SitemapListener.php]: this method was not meant to be used here. We could use `ModuleNews::sortOutProtected()` instead (maybe make it `public static`) or use the security voter instead.

File: news-bundle/src/EventListener/SitemapListener.php [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+            $targetPage->loadDetails();
+
+            // Skip news outside the current page root IDs
+            if (!\in_array((int) $targetPage->rootId, $rootPageIds, true)) {
+                continue;
+            }
+
+            // Skip target page if protected and cannot be accessed
+            if (!$controllerAdapter->isVisibleElement($targetPage)) {
```
Comment (aschempp) [inline news-bundle/src/EventListener/SitemapListener.php]: don't reuse this method here. use the security voter instead

File: news-bundle/src/EventListener/SitemapListener.php (hunk 1 above)
Comment (fritzmg) [reply news-bundle/src/EventListener/SitemapListener.php]: What do you mean? We are using annotations for service tagging in countless services already. But yes, I can change it to PHP 8 attributes instead. A separate PR should then may be change _all_ the annotations to attributes for consistency.

File: news-bundle/src/EventListener/SitemapListener.php (hunk 2 above)
Comment (fritzmg) [reply news-bundle/src/EventListener/SitemapListener.php]: True, it's just my IDE that's unable to parse that.

File: news-bundle/src/EventListener/SitemapListener.php (hunk 3 above)
Comment (fritzmg) [reply news-bundle/src/EventListener/SitemapListener.php]: > this method was not meant to be used here.

Why not? `Controller::isVisibleElement` is not restricted to any particular model. It is applied, and can be applied, to any model that contains the following fields:

* `protected`
* `groups`

Currently it is applied to instances of `ArticleModel`, `ContentModel` and `ModuleModel` and I do not see any reason not to apply it here as well, as it already covers everything that is necessary and also implements the `isVisibleElement` hook, which is still not deprecated.

We might of course decide to deprecate this method and its hook - but that is unrelated to this PR and should be done in a separate PR. This PR can then be updated accordingly before or after the fact.

File: news-bundle/src/EventListener/SitemapListener.php (hunk 4 above)
Comment (fritzmg) [reply news-bundle/src/EventListener/SitemapListener.php]: See above.

**Comment by leofeyer** [comment]
Now that #4548 has been merged, we should probably align this PR with it.

**Comment by fritzmg** [comment]
> Now that #4548 has been merged, we should probably align this PR with it.

#4548 does not include protected calenders in the feed. Since the feed is generated by a Controller in 4.9+ we now have the opportunity to also include protected items in the feed. The current state of my PR would allow this (although my usage of `Controller::isVisibleElement` to achieve this without code duplication is contested by @aschempp).

**Comment by Toflar** [comment]
Let's migrate to the event first and add new features later.

File: news-bundle/tests/EventListener/SitemapListenerTest.php [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        ];
+
+        $sitemapEvent = $this->createSitemapEvent([]);
+        $listener = $this->createListener([], $adapters);
+        $listener($sitemapEvent);
+
+        $this->assertStringNotContainsString('<url><loc>', (string) $sitemapEvent->getDocument()->saveXML());
+    }
+
+    public function testCalendarEventIsAdded(): void
```
Comment (Toflar) [inline news-bundle/tests/EventListener/SitemapListenerTest.php]: A bit too much copy&paste ;)

Comment (fritzmg) [reply news-bundle/tests/EventListener/SitemapListenerTest.php]: 🙈

**Review by Toflar — DISMISSED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @fritzmg.

**Comment by leofeyer** [comment]
Thank you @fritzmg.

**Pull request merged**